sonar-batch-protocol/src/main/java/org/sonar/batch/protocol/viewer/ViewerApplication.java

```diff
@@ -0,0 +1,341 @@
+import javax.swing.tree.DefaultMutableTreeNode;
+import javax.swing.tree.DefaultTreeModel;
+import javax.swing.tree.TreeSelectionModel;
+import org.sonar.batch.protocol.output.BatchReport;
+import org.sonar.batch.protocol.output.BatchReport.Component;
+import org.sonar.batch.protocol.output.BatchReport.Metadata;
+import org.sonar.batch.protocol.output.BatchReportReader;
+import org.sonar.batch.protocol.output.FileStructure.Domain;
+import org.sonar.core.util.CloseableIterator;
+
+public class ViewerApplication {
+
+  private JFrame frame;
+  private BatchReportReader reader;
+  private Metadata metadata;
+  private SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy HH:mm:ss");
+  private JTree componentTree;
+  private JSplitPane splitPane;
+  private JTabbedPane tabbedPane;
+  private JScrollPane treeScrollPane;
+  private JScrollPane componentDetailsTab;
+  private JScrollPane highlightingTab;
+  private JEditorPane componentEditor;
+  private JEditorPane highlightingEditor;
+  private JScrollPane sourceTab;
+  private JEditorPane sourceEditor;
+  private JScrollPane coverageTab;
+  private JEditorPane coverageEditor;
+  private TextLineNumber textLineNumber;
+
+  /**
+   * Launch the application.
+   */
+  public static void main(String[] args) {
+    EventQueue.invokeLater(new Runnable() {
+      public void run() {
+        try {
+          ViewerApplication window = new ViewerApplication();
+          window.frame.setVisible(true);
+
+          window.loadReport();
+        } catch (Exception e) {
+          e.printStackTrace();
+        }
+      }
+
+    });
+  }
+
+  private void loadReport() {
+    final JFileChooser fc = new JFileChooser();
+    fc.setDialogTitle("Choose batch report directory");
+    fc.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY);
+    fc.setApproveButtonText("Open batch report");
+    int returnVal = fc.showOpenDialog(frame);
+    if (returnVal == JFileChooser.APPROVE_OPTION) {
+      File file = fc.getSelectedFile();
+      try {
+        loadReport(file);
+      } catch (Exception e) {
+        JOptionPane.showMessageDialog(frame, e.getMessage(), "Error", JOptionPane.ERROR_MESSAGE);
+        exit();
+      }
+    } else {
+      exit();
+    }
+
+  }
+
+  private void exit() {
+    frame.setVisible(false);
+    frame.dispose();
+  }
+
+  private void loadReport(File file) {
+    reader = new BatchReportReader(file);
+    metadata = reader.readMetadata();
+    updateTitle();
+    loadComponents();
+  }
+
+  private void loadComponents() {
+    int rootComponentRef = metadata.getRootComponentRef();
+    Component component = reader.readComponent(rootComponentRef);
+    DefaultMutableTreeNode project = createNode(component);
+    loadChildren(component, project);
+    getComponentTree().setModel(new DefaultTreeModel(project));
+  }
+
+  private DefaultMutableTreeNode createNode(Component component) {
+    DefaultMutableTreeNode node = new DefaultMutableTreeNode(component) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Immediately return this expression instead of assigning it to the temporary variable "node". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1488)